### ISSUE TITLE
Support for running commands and subscribing to events for multiple user contexts

### ISSUE THREAD

**Issue by whimboo** [w3c/webdriver-bidi#775]
We should discuss the possibility of enhancing WebDriver BiDi to support running commands and subscribing to events for a specified list of user contexts. This would provide more granular control over how commands are issued and events are captured in multi-user context scenarios.

Here some examples:

1. Use the `browsingContext.setViewport` command to set the viewport for all the tabs from user context 1 at the same time. Same could be imagined for a future user agent emulation.
2. Subscribe to events like `browsingContext.contextCreated` only for tabs of a given user context.

**Comment by whimboo** [comment]
Configuration options as used by Playwright ([channels.d.ts](https://github.com/microsoft/playwright/blob/ada68cd6f0cf41b4788a79b8faac44c08cebfe68/packages/protocol/src/channels.d.ts#L1226C13-L1226C37)) are the following:

- [ ] noDefaultViewport
- [x] viewport ([PR](https://github.com/w3c/webdriver-bidi/pull/876))
- [x] screen ([PR](https://github.com/w3c/webdriver-bidi/pull/925))
- [x] ignoreHTTPSErrors ([PR](https://github.com/w3c/webdriver-bidi/pull/895))
- [x] clientCertificates ([PR](https://github.com/w3c/webdriver-bidi/pull/895))
- [x] javaScriptEnabled
- [ ] bypassCSP
- [x] userAgent
- [x] locale ([PR](https://github.com/w3c/webdriver-bidi/pull/924))
- [x] timezoneId
- [x] geolocation ([PR](https://github.com/w3c/webdriver-bidi/pull/888))
- [ ] permissions ([PR](https://github.com/w3c/permissions/issues/437) [permission spec])
- [x] extraHTTPHeaders
- [x] offline
- [ ] httpCredentials
- [x] deviceScaleFactor ([PR](https://github.com/w3c/webdriver-bidi/pull/876))
- [ ] isMobile
- [ ] hasTouch
- [x] colorScheme
- [ ] reducedMotion
- [x] forcedColors ([PR](https://github.com/w3c/webdriver-bidi/pull/905))
- [x] acceptDownloads ([PR](https://github.com/w3c/webdriver-bidi/pull/883) + [PR](https://github.com/w3c/webdriver-bidi/pull/923))
- [ ] baseURL 
- [ ] recordVideo
- [ ] recordHar
- [ ] strictSelectors
- [ ] serviceWorkers
- [x] proxy ([PR](https://github.com/w3c/webdriver-bidi/pull/897))
- [ ] storageState

**Comment by juliandescottes** [comment]
@yury-s we have either added options to user context creation or updated commands to target user contexts, so we are now able to setup:
- proxy
- allow insecure certificates
- viewport
- geolocation
- events

Are there other APIs which are critical for Playwright to be able to set on the user context level?

**Comment by yury-s** [comment]
Looking at the parameters of [browser.newContext()](https://playwright.dev/docs/api/class-browser#browser-new-context) and the list [above](https://github.com/w3c/webdriver-bidi/issues/775#issuecomment-2577785800), I can immediately spot several critical ones that require browser support: bypassCSP, isMobile, javaScriptEnabled, locale, timezoneId, and offline.

Our test suite assumes full support for them, and we use it as the ultimate compatibility check. You can treat it as a de facto acceptance suite for spec completeness - if the suite passes, support is likely sufficient.

**Comment by juliandescottes** [comment]
Thanks. Trying to link this to existing issues
- isMobile: https://github.com/w3c/webdriver-bidi/issues/772 (part of setViewport, so should be fine)
- offline: needs to be handled in https://github.com/w3c/webdriver-bidi/issues/776 (throttling, new command, need to add userContext support)
- javaScriptEnabled: https://github.com/w3c/webdriver-bidi/issues/746 (probably a new command, no clear decision on the issue)
- locale: https://github.com/w3c/webdriver-bidi/issues/774 & https://github.com/w3c/webdriver-bidi/pull/924
- timezone: https://github.com/w3c/webdriver-bidi/issues/749

@yury-s For `bypassCSP` is it something you expose and allow to set to false, or do you always want to set it to true? With the current spec, I think the assumption is that all code executed via script.evaluate or script.callFunction should not be impacted by CSPs. We actually have https://bugzilla.mozilla.org/show_bug.cgi?id=1941780 to fix it on Firefox's side

**Comment by whimboo** [comment]
@yury-s maybe you can find some minutes to answer the question from @juliandescottes around `bypassCSP`? Thanks!

**Comment by yury-s** [comment]
Sorry for the late response, it somehow fell through the cracks. 

>  For bypassCSP is it something you expose and allow to set to false, or do you always want to set it to true?

The idea is to allow users to set it `true` to disable default CSP behavior and allow running _user_ scripts that otherwise would be blocked, see `browsercontext-csp.spec.ts`. The examples include adding `<script>` tag, but this is mostly for the user code to not fail on the checks, so it should be opt-in.

> With the current spec, I think the assumption is that all code executed via script.evaluate or script.callFunction should not be impacted by CSPs. We actually have https://bugzilla.mozilla.org/show_bug.cgi?id=1941780 to fix it on Firefox's side

This matches Playwright behavior, `page.evaluate()` should work even when CSP is enforced, see CSP tests in `page-evaluate.spec.ts`.

**Comment by juliandescottes** [comment]
> Sorry for the late response, it somehow fell through the cracks.
> 
> > For bypassCSP is it something you expose and allow to set to false, or do you always want to set it to true?
> 
> The idea is to allow users to set it `true` to disable default CSP behavior and allow running _user_ scripts that otherwise would be blocked, see `browsercontext-csp.spec.ts`. The examples include adding `<script>` tag, but this is mostly for the user code to not fail on the checks, so it should be opt-in.
> 
> > With the current spec, I think the assumption is that all code executed via script.evaluate or script.callFunction should not be impacted by CSPs. We actually have https://bugzilla.mozilla.org/show_bug.cgi?id=1941780 to fix it on Firefox's side
> 
> This matches Playwright behavior, `page.evaluate()` should work even when CSP is enforced, see CSP tests in `page-evaluate.spec.ts`.

Note that script.evaluate/callFunction already work in Firefox when CSP is enforced, but using new Function / eval in those scripts is not supported. So from the CSP tests in `page.evaluate.spec.ts`, `should work with CSP` already passes, but `should work with new Function() and CSP` doesn't.

@yury-s  So in theory, if I use `bypassCSP: true` on a context, does it remove all restrictions related to CSP? If a script loaded by the test page (not via page.evaluate, or addScriptTag) uses eval/new Function does it work then?
I'm not sure I understand the added value compared to evaluating scripts directly with the dedicated script commands. In which case is it helpful to bypass CSP to run a user script which can't be handled via script.evaluate/callFunction?

**Comment by yury-s** [comment]
> [@yury-s](https://github.com/yury-s) So in theory, if I use `bypassCSP: true` on a context, does it remove all restrictions related to CSP? If a script loaded by the test page (not via page.evaluate, or addScriptTag) uses eval/new Function does it work then?

Yes, the idea is that it removes all restrictions related to CSP. E.g. the following page should load and run the inline scripts just fine (I'm surprised we don't have a test covering that):

```js
<meta http-equiv="Content-Security-Policy" content="default-src 'self'">
<script>
  console.log('logging from csp.html');
  new Function('console.log("new function");')();
</script>
```
